File: src/worker/core/season/addDaysToSchedule.cs
// Framework: .NET

using System;
using System.Collections.Generic;
using System.Linq;

public class Game
{
    public int Season { get; set; }
    public int Day { get; set; }
}

public class ScheduleGameWithoutKey
{
    public int HomeTid { get; set; }
    public int AwayTid { get; set; }
    public int Day { get; set; }
}

public static class Util
{
    public static Dictionary<string, object> g = new Dictionary<string, object>
    {
        { "season", 2023 } // Example season value
    };
}

public static class ScheduleManager
{
    public static List<ScheduleGameWithoutKey> AddDaysToSchedule(
        List<(int HomeTid, int AwayTid)> games,
        List<Game> existingGames = null)
    {
        var dayTids = new HashSet<int>();
        bool prevDayAllStarGame = false;
        bool prevDayTradeDeadline = false;

        int day = 1;

        // If there are other games already played this season, start after that day
        if (existingGames != null)
        {
            int season = (int)Util.g["season"];
            foreach (var game in existingGames)
            {
                if (game.Season == season && game.Day >= day)
                {
                    day = game.Day + 1;
                }
            }
        }

        return games.Select(game =>
        {
            int awayTid = game.AwayTid;
            int homeTid = game.HomeTid;

            bool allStarGame = awayTid == -2 && homeTid == -1;
            bool tradeDeadline = awayTid == -3 && homeTid == -3;
            if (dayTids.Contains(homeTid) ||
                dayTids.Contains(awayTid) ||
                allStarGame ||
                prevDayAllStarGame ||
                tradeDeadline ||
                prevDayTradeDeadline)
            {
                day += 1;
                dayTids.Clear();
            }

            dayTids.Add(homeTid);
            dayTids.Add(awayTid);

            prevDayAllStarGame = allStarGame;
            prevDayTradeDeadline = tradeDeadline;

            return new ScheduleGameWithoutKey
            {
                HomeTid = homeTid,
                AwayTid = awayTid,
                Day = day
            };
        }).ToList();
    }
}

File: src/worker/core/season/addDaystoSchedule.f
module GameSchedule
  implicit none
contains

  function addDaysToSchedule(games, existingGames) result(scheduledGames)
    type :: Game
      integer :: homeTid
      integer :: awayTid
      integer :: day
    end type Game

    type :: ScheduleGameWithoutKey
      integer :: homeTid
      integer :: awayTid
      integer :: day
    end type ScheduleGameWithoutKey

    type(Game), dimension(:), allocatable :: games
    type(Game), dimension(:), optional :: existingGames
    type(ScheduleGameWithoutKey), dimension(:), allocatable :: scheduledGames

    integer :: dayTids(100)  ! Assuming a maximum of 100 teams
    logical :: prevDayAllStarGame
    logical :: prevDayTradeDeadline
    integer :: day
    integer :: i, season

    prevDayAllStarGame = .false.
    prevDayTradeDeadline = .false.
    day = 1

    ! If there are other games already played this season, start after that day
    if (present(existingGames)) then
      season = g%get("season")
      do i = 1, size(existingGames)
        if (existingGames(i)%season == season .and. existingGames(i)%day >= day) then
          day = existingGames(i)%day + 1
        end if
      end do
    end if

    allocate(scheduledGames(size(games)))
    do i = 1, size(games)
      integer :: awayTid, homeTid
      logical :: allStarGame, tradeDeadline

      awayTid = games(i)%awayTid
      homeTid = games(i)%homeTid

      allStarGame = (awayTid == -2 .and. homeTid == -1)
      tradeDeadline = (awayTid == -3 .and. homeTid == -3)

      if (any(dayTids == homeTid) .or. any(dayTids == awayTid) .or. &
          allStarGame .or. prevDayAllStarGame .or. &
          tradeDeadline .or. prevDayTradeDeadline) then
        day = day + 1
        dayTids = 0  ! Clear the dayTids array
      end if

      dayTids(i) = homeTid
      dayTids(i+1) = awayTid

      prevDayAllStarGame = allStarGame
      prevDayTradeDeadline = tradeDeadline

      scheduledGames(i)%homeTid = homeTid
      scheduledGames(i)%awayTid = awayTid
      scheduledGames(i)%day = day
    end do
  end function addDaysToSchedule

end module GameSchedule
